File: app/src/ipc/bookmark.ts
import { ipcInvoke } from "@/ipc/client";
import type { Bookmark } from "@/ipc/types";

/**
 * `bookmark.*` namespace (§43.1). Mirrors the new backend `bookmark_*`
 * Tauri commands, thin passthroughs to the pre-existing
 * `BookmarkRepository` (§33.9, `atlas-memory`/`atlas-db`).
 */
export function bookmarkList(documentId: number): Promise<Bookmark[]> {
  return ipcInvoke<Bookmark[]>("bookmark_list", { documentId });
}

export function bookmarkCreate(documentId: number, locationRef: string, label: string): Promise<Bookmark> {
  return ipcInvoke<Bookmark>("bookmark_create", { documentId, locationRef, label });
}

export function bookmarkDelete(bookmarkId: number): Promise<void> {
  return ipcInvoke<void>("bookmark_delete", { bookmarkId });
}

File: app/src-tauri/crates/app-tauri/src/commands/bookmark.rs
//! `bookmark.*` namespace (§43.1, new for the Document Experience
//! milestone): thin passthrough to the already-implemented
//! `BookmarkRepository` (§33.9), reached via `AppFacade::memory_engine`.
//! Handlers only validate/forward/map errors (§26, §46.4).

use tauri::State;

use atlas_core::AppFacade;
use atlas_types::ids::{BookmarkId, DocumentId};
use atlas_types::memory::Bookmark;
use atlas_utils::time::now_iso8601;
use atlas_utils::AppError;

#[tauri::command]
pub fn bookmark_list(
    facade: State<'_, AppFacade>,
    document_id: i64,
) -> Result<Vec<Bookmark>, AppError> {
    facade
        .memory_engine()
        .bookmarks()
        .list_for_document(DocumentId(document_id))
}

#[tauri::command]
pub fn bookmark_create(
    facade: State<'_, AppFacade>,
    document_id: i64,
    location_ref: String,
    label: String,
) -> Result<Bookmark, AppError> {
    facade.memory_engine().bookmarks().insert(Bookmark {
        id: BookmarkId(0),
        document_id: DocumentId(document_id),
        location_ref,
        label,
        created_at: now_iso8601(),
    })
}

#[tauri::command]
pub fn bookmark_delete(
    facade: State<'_, AppFacade>,
    bookmark_id: i64,
) -> Result<(), AppError> {
    facade.memory_engine().bookmarks().delete(BookmarkId(bookmark_id))
}
